DBZ-1557 DB2 Plugin Documentation (#1204)
debezium/debezium: documentation/modules/ROOT/pages/connectors/db2.adoc

= Debezium Connector for Db2 
include::../_attributes.adoc[]
:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

> [NOTE]

Reviewer: Can you please add the same note on the incubating status [as we have](https://github.com/debezium/debezium/blob/master/documentation/modules/ROOT/pages/connectors/oracle.adoc) for the Oracle connector?

Author: Done

====
Want to help us further hone and improve it? link:/docs/contribute/[Learn how].
====

toc::[]

Debezium's Db2 Connector can monitor and record the row-level changes in the tables of a Db2 database.
This connector was added in release Debezium 1.1.x.
It is strongly inspired by the Debezium implementation of  SQL server, using a

> SQL based polling model of tables put into "capture mode". This uses the ASN libraries that come as a standard part of Db2 LUW and can be added (with an additional license) to Db2 zOS.

Reviewer: Needs update as per the licensing discussion.

Author: Do you have a "boilerplate" text for this ?

Reviewer: There's something similar for Oracle: https://github.com/debezium/debezium/blob/master/documentation/modules/ROOT/pages/connectors/oracle.adoc#deploying-a-connector

For DB2 this should not recommend to copy the driver to Connect lib dir, but rather the connector directory itself.

Author: Are you referring here specifically to the license for jdbc OR for the usage of the Db2 ASN services ?

Reviewer: Here I'm referring to the IIDR license. I'm going to push a commit for this to this PR.

The first time it connects to a Db2 database, it reads a consistent snapshot of all of the tables that are whitelisted (or not blacklisted depending on the mode of operation).
Note that by default all tables in the database are snapshoted, NOT just those that are in capture mode.

When that snapshot is complete, the connector continuously streams the changes that were committed to the Db2 database for all whitelisted tables in capture mode. This generates corresponding insert, update and delete events.
All of the events for each table are recorded in a separate Kafka topic, where they can be easily consumed by applications and services.

[[overview]]
== Overview

The functionality of the connector is based upon the ASN Capture/Apply agents 
that enable SQL Replication within DB2:

> https://www.ibm.com/support/pages/q-replication-and-sql-replication-product-documentation-pdf-format-version-101-linux-unix-and-windows

Reviewer: Should be a link with proper text, e.g. in the "ASN Capture/Apply agents" part in the sentence.

Author: Done

The capture agent generates ASN tables for all tables in CDC mode, these ASN tables can be queried via a SQL interface in order to read change events.
Using this mechanism a Db2 ASN capture agent monitors all databases and tables that the user is interested in and stores the changes into specifically created ASN tables.

> The connector has been tested with Db2/Linux 11.5.0.0, but our expectation is that the model would work for Windows, AIX and zOS as wll.

Reviewer: ```suggestion
The connector has been tested with Db2/Linux 11.5.0.0, but our expectation is that the model would work for Windows, AIX and zOS as well.
```

Author: Done

[[setting-up-Db2]]
== Setting up Db2 

Before using the Db2 connector to monitor the changes committed on the Db2 database, first enable _CDC_ on a monitored database.

We recommend using the UDF functions that we provide.
However these are only for convenience and manually performing the same commands using
db2 control commands would have the same effect. 

We assume here that the contents of 
debezium-connector-db2/src/test/docker/db2-cdc-docker are available
on the machine on which Db2 is running in $HOME/asncdctools/src
and that the user has logged-in as the db2inst1 user.

First compile this function on the Db2 server, using the _bldrtn_
C-compiler

[source,shell]

----
$ cd $HOME/asncdctools/src
----

[source,shell]

----
$ ./bldrtn asncdc
----

Start the database if not already running:

[source,shell]

----
$ db2 start db <DB NAME>

----

We need to make sure that the metadata catalog can be read via JDBC

[source,shell]
----
$ cd $HOME/sqllib/bnd
----

[source,shell]
----
$ db2 bind db2schema.bnd blocking all grant public sqlerror continue 
----

The database needs to have been recently backed-up so doing following if this is not the case:

[source,shell]
----

> $ db2 backup db <DB NAME> to /dev/null

Reviewer: Having no experience with that tool: what does the "to /dev/null" apply to? Any log messages, or is it sort of a joke as target for the backup? In that case I'd prefer to remove it so to avoid any surprises by unknowing users :)

Author: I added some more explanation in the text

----

[source,shell]
----
$ db2 restart db <DB NAME>
----

Install the UDF:

[source,shell]

----
$ db2 connect to <DB NAME>
----

> We assume that the db2 is installed on the db2inst1 user

Reviewer: Is it "db2" or "Db2"? I'm seeing both, would be good to converge on one.

Author: It is Db2 for the Database, but the the command to connect to it is _db2_ I have tried to
use consistently.
Done.

[source,shell]

----
$ db2 -tvmf $HOME/asncdctools/src/asncdcaddremove.sql
----

Having done the above, we can use the UDFs to control ASN via SQL commands.
Some of the UDFs expect a return value in which case we use the SQL _VALUE_
statement to invoke them, while others are fire-and-forget in which case we
use the SQL _CALL_ statement.

First the ASN agent needs to be started:

[source,sql]
----
VALUES ASNCDC.ASNCDCSERVICES('start','asncdc');
----

If the agent every needs to be stopped:

[source,sql]
----
VALUES ASNCDC.ASNCDCSERVICES('stop','asncdc');
----

The status if the agent can be checked at any moment:

[source,sql]
----
VALUES ASNCDC.ASNCDCSERVICES('status','asncdc');
----

A table MYTABLE in MYSCHEMA can be put into capture mode by doing:

[source,sql]
----
CALL ASNCDC.ADDTABLE('MYSCHEMA', 'MYTABLE' )";
----

At table MYTABLE in MYSCHEMA can be removed from capture mode by doing:

[source,sql]
----
CALL ASNCDC.REMOVETABLE('MYSCHEMA', 'MYTABLE' );
----

After a table is added or removed, the user __MUST__ reinitialize the ASN service:

[source,sql]
----
VALUES ASNCDC.ASNCDCSERVICES('reinit','asncdc');
----

> == How the Db2 connector works

Reviewer: Is there mention of downloading the JDBC driver and adding it to the contents of the connector archive somewhere?

Author: We have a dependency in the pom file.
We don't explicitly download the jdbc driver.
We assume the release process when the generating the jars pulls the dependencies.
What do we need to add in the documentation ?

Reviewer: Debezium is typically not consumed as a Maven dependency (the [embedded mode](https://debezium.io/documentation/reference/1.0/operations/embedded.html) being the exception)), but as a plug-in that's added to Kafka Connect. Those plug-ins are obtained as ZIP/TAR.GZ archives from Maven Central. In case of DB2, we cannot add the driver JAR to that connector archive due to its license (unlike as for most other connectors), so a user must

* obtain and extract the DB2 connector archive
* obtain the JDBC driver and add it into the extraced connector directory

That's what I started the concurrent discussion with Erin for, trying to gauge whether we actually can have this changed so we can redistribute the driver in the connector archive.

Author: Ok. Sorry I misunderstood what you meant.
Done.

4. Store the maximum LSN and repeat the loop.

After a restart, the connector will resume from the offset (commit and change LSNs) where it left off before.

The connector is able to detect whether the CDC is enabled or disabled for whitelisted source table during the runtime and modify its behaviour.

=== Topic names

The Db2 connector writes events for all insert, update, and delete operations on a single table to a single Kafka topic. The name of the Kafka topics always takes the form _databaseName_._schemaName_._tableName_, where _databaseName_ is the logical name of the connector as specified with the `database.server.name` configuration property, _schemaName_ is the name of the schema where the operation occurred, and _tableName_ is the name of the database table on which the operation occurred.

Unlike for SQL Server, it is only possible for a connector to streams changes from one Db2 database.

For example, consider a Db2 installation with an `mydatabase` database that contains four tables: `products`, `products_on_hand`, `customers`, and `orders` in schema `myschema` then the connector would produce events on these four Kafka topics:

> * `mydatabase.myschema.products`

Reviewer: Schema and table name should be upper-cased as per the recent discussion with @lga-zurich.

Author: Done

The schema describes the structure of the payload, while the payload contains the actual data.

[[change-event-keys]]
==== Change Event Keys

For a given table, the change event's key will have a structure that contains a field for each column in the primary key (or unique key constraint) of the table at the time the event was created.

Consider a `customers` table defined in the `mydatabase` database's schema `myschema`:

[source,sql,indent=0]
----
CREATE TABLE customers (
  id INTEGER IDENTITY(1001,1) NOT NULL PRIMARY KEY,
  first_name VARCHAR(255) NOT NULL,
  last_name VARCHAR(255) NOT NULL,
  email VARCHAR(255) NOT NULL UNIQUE
);
----

The `database.server.name` configuration property is then `mydatabase`.
Every change event for the `customers` table while it has this definition will feature the same key structure, which in JSON looks like this:

[source,json,indent=0]
----
{
    "schema": {
        "type": "struct",
        "fields": [
            {
                "type": "int32",
                "optional": false,
                "field": "id"
            }
        ],
        "optional": false,
        "name": "mydatabase.myschema.customers.Key"
    },
    "payload": {

>         "id": 1004

Reviewer: Field names are upper-case, I think?

          },
          {
            "type": "string",
            "optional": false,
            "field": "db"
          },
          {
            "type": "string",
            "optional": false,
            "field": "schema"
          },
          {
            "type": "string",
            "optional": false,
            "field": "table"
          },
          {
            "type": "string",
            "optional": true,
            "field": "change_lsn"
          },
          {
            "type": "string",
            "optional": true,
            "field": "commit_lsn"
          },
        ],
        "optional": false,
        "name": "io.debezium.connector.db2.Source",
        "field": "source"
      },
      {
        "type": "string",
        "optional": false,
        "field": "op"
      },
      {
        "type": "int64",
        "optional": true,
        "field": "ts_ms"
      }
    ],
    "optional": false,
    "name": "mydatabase.myschema.customers.Envelope"
  },
  "payload": {
    "before": null,
    "after": {
      "id": 1005,
      "first_name": "john",
      "last_name": "doe",
      "email": "john.doe@example.org"
    },
    "source": {

>       "version": "0.10.0.Alpha1",

Reviewer: ```suggestion
      "version": "{debezium-version}",
```

Author: Done

      "connector": "db2",
      "name": "myconnector",
      "ts_ms": 1559729468470,
      "snapshot": false,
      "db": "mydatabase",
      "schema": "myschema",

>       "table": "customers",

Reviewer: Same note on casing. Would be good to compare to an actual record from the connector.

[TIP]
====
It may appear that the JSON representations of the events are much larger than the rows they describe.
This is true, because the JSON representation must include the _schema_ and the _payload_ portions of the message.
It is possible and even recommended to use the link:/docs/faq/#avro-converter[Avro Converter] to dramatically decrease the size of the actual messages written to the Kafka topics.
====

[[update-events]]
===== Update events
The value of an _update_ change event on this table will actually have the exact same _schema_, and its payload will be structured the same but will hold different values.
Here's an example:

[source,json,indent=0,subs="attributes"]
----
{
  "schema": { ... },
  "payload": {
    "before": {
      "id": 1005,
      "first_name": "john",
      "last_name": "doe",
      "email": "john.doe@example.org"
    },
    "after": {
      "id": 1005,
      "first_name": "john",
      "last_name": "doe",
      "email": "noreply@example.org"
    },
    "source": {

>       "version": "0.10.0.Alpha1",

Reviewer: ```suggestion
      "version": "{debezium-version}",
```

Author: Done

The `source` structure tells us information about Db2's record of this change (providing traceability), but more importantly this has information we can compare to other events in this and other topics to know whether this event occurred before, after, or as part of the same Db2 commit as other events.

[NOTE]
====
When the columns for a row's primary/unique key are updated, the value of the row's key has changed so Debezium will output _three_ events: a `DELETE` event and a link:#tombstone-events[tombstone event] with the old key for the row, followed by an `INSERT` event with the new key for the row.
====

[[delete-events]]
===== Delete events

So far we've seen samples of _create_ and _update_ events.
Now, let's look at the value of a _delete_ event for the same table. Once again, the `schema` portion of the value will be exactly the same as with the _create_ and _update_ events:

[source,json,indent=0,subs="attributes"]
----
{
  "schema": { ... },
  },
  "payload": {
    "before": {
      "id": 1005,
      "first_name": "john",
      "last_name": "doe",
      "email": "noreply@example.org"
    },
    "after": null,
    "source": {

>       "version": "0.10.0.Alpha1",

Reviewer: ```suggestion
      "version": "{debezium-version}",
```

Author: Done.

The procedure itself is also much simpler than in case of cold schema update.
First we consider an incremental change to the source, e.g. adding a new column to the end:

1. lock the source table to change
2.  Mark the tables as INACTIVE on the ASN register table and reinit the ASN capture service (see the UDFs)34. Apply all changes to the source table schema
3.  Apply all changes to the ASN table schema
4.  Mark the tables as ACTIVE on the ASN register table and reinit the ASN capture service (see the UDFs)

Now we consider an non-incremental change to the source, e.g. adding a new column in the middle:

1. lock the source table to change
2.  Mark the tables as INACTIVE on the ASN register table and reinit the ASN capture service (see the UDFs)
3. export the data of the source table to change
4. truncate the source table
5. alter the source table
6. LOAD the exported data into the altered source table 
7. export the data of the ASN table to change
8. truncate the ASN table
9. alter the ASN table
10. LOAD the exported data into the altered ASN table 
11. Mark the tables as INACTIVE on the ASN register table and reinit the ASN capture service (see the UDFs)

==== Example

> To Be Done

Reviewer: Are you planning to do this as part of this PR?

Author: No I don't think we will have time,

Reviewer: Ok, I've added this to the list of follow-up tasks in [DBZ-1752](https://issues.redhat.com/browse/DBZ-1752).

[[data-types]]
=== Data types

> https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0008483.html

Reviewer: Should have a link text.

Author: Done

|`DECIMAL[(P[,S])]`
|`BYTES`
|`org.apache.kafka.connect.data.Decimal`
|The `scale` schema parameter contains an integer representing how many digits the decimal point was shifted.
The `connect.decimal.precision` schema parameter contains an integer representing the precision of the given decimal value.

|`SMALLMONEY`
|`BYTES`
|`org.apache.kafka.connect.data.Decimal`
|The `scale` schema parameter contains an integer representing how many digits the decimal point was shifted.
The `connect.decimal.precision` schema parameter contains an integer representing the precision of the given decimal value.

|`MONEY`
|`BYTES`
|`org.apache.kafka.connect.data.Decimal`
|The `scale` schema parameter contains an integer representing how many digits the decimal point was shifted.
The `connect.decimal.precision` schema parameter contains an integer representing the precision of the given decimal value.

|=======================

[[deploying-a-connector]]
== Deploying a connector

If you've already installed https://zookeeper.apache.org[Zookeeper], http://kafka.apache.org/[Kafka], and http://kafka.apache.org/documentation.html#connect[Kafka Connect], then using Debezium's Db2` connector is easy.

> Simply download the https://repo1.maven.org/maven2/io/debezium/debezium-connector-sqlserver/0.9.0.Alpha1/debezium-connector-sqlserver-0.9.0.Alpha1-plugin.tar.gz[connector's plugin archive], extract the JARs into your Kafka Connect environment, and add the directory with the JARs to http://docs.confluent.io/{confluent-platform-version}/connect/userguide.html#installing-plugins[Kafka Connect's classpath].

Reviewer: Please use {debezium-version} for the version. Also /sqlserver/db2/.

Author: Done

> Simply download the https://repo1.maven.org/maven2/io/debezium/debezium-connector-sqlserver/0.9.0.Alpha1/debezium-connector-sqlserver-0.9.0.Alpha1-plugin.tar.gz[connector's plugin archive], extract the JARs into your Kafka Connect environment, and add the directory with the JARs to http://docs.confluent.io/{confluent-platform-version}/connect/userguide.html#installing-plugins[Kafka Connect's classpath].

Reviewer: So this also answers one of my questions above; here the procedure for downloading and adding the JDBC driver should be described.

Author: See previous comments.

Restart your Kafka Connect process to pick up the new JARs.

> If immutable containers are your thing, then check out https://hub.docker.com/r/debezium/[Debezium's Docker images] for Zookeeper, Kafka and Kafka Connect with the Db2 connector already pre-installed and ready to go.

Reviewer: ```suggestion
If immutable containers are your thing, then check out https://hub.docker.com/r/debezium/[Debezium's container images] for Zookeeper, Kafka and Kafka Connect with the Db2 connector already pre-installed and ready to go.
```

We generally prefer "container images", as Docker is just one particular implementation of a container runtime.

Author: Done.

You can even link:/docs/openshift/[run Debezium on OpenShift].

To use the connector to produce change events for a particular Db2 database or cluster:

. enable the link:#setting-up-Db2[CDC on Db2] to publish the _CDC_ events in the database
. create a link:#example-configuration[configuration file for the Db2 Connector] and use the https://docs.confluent.io/{confluent-platform-version}/connect/restapi.html[Kafka Connect REST API] to add that connector to your Kafka Connect cluster.

When the connector starts, it will grab a consistent snapshot of the schemas in your Db2 database and start streaming changes, producing events for every inserted, updated, and deleted row.
You can also choose to produce events for a subset of the schemas and tables.
Optionally ignore, mask, or truncate columns that are sensitive, too large, or not needed.

[[example]]
[[example-configuration]]
=== Example configuration

> Using the Db2 connector is straightforward. Here is an example of the configuration for a connector instance that monitors a Db2 server at port 3306 on 192.168.99.100, which we logically name `fullfillment`:

Reviewer: ```suggestion
Using the Db2 connector is straightforward. Here is an example of the configuration for a connector instance that monitors a Db2 server at port 50000 on 192.168.99.100, which we logically name `fullfillment`:
```

As per the example in the code.

<11> The name of the database history topic where the connector will write and recover DDL statements. This topic is for internal use only and should not be used by consumers.

See the link:#connector-properties[complete list of connector properties] that can be specified in these configurations.

This configuration can be sent via POST to a running Kafka Connect service, which will then record the configuration and start up the one connector task that will connect to the Db2 database, read the transaction log, and record events to Kafka topics.

[[monitoring]]
=== Monitoring

Kafka, Zookeeper, and Kafka Connect all have xref:operations/monitoring.adoc[built-in support] for JMX metrics. The Db2 connector also publishes a number of metrics about the connector's activities that can be monitored through JMX. The connector has two types of metrics. Snapshot metrics help you monitor the snapshot activity and are available when the connector is performing a snapshot. Streaming metrics help you monitor the progress and activity while the connector reads CDC table data.

[[monitoring-snapshots]]
[[snapshot-metrics]]
==== Snapshot Metrics

> ===== *MBean: debezium.sql_server:type=connector-metrics,context=snapshot,server=_<database.server.name>_*

Reviewer: sql_server -> db2

Author: Done

|`string[]`
|The list of tables that are monitored by the connector.

|`QueueTotalCapcity`
|`int`
|The length of the queue used to pass events between the snapshotter and the main Kafka Connect loop.

|`QueueRemainingCapcity`
|`int`
|The free capacity of the queue used to pass events between the snapshotter and the main Kafka Connect loop.

|`TotalTableCount`
|`int`
|The total number of tables that are being included in the snapshot.

|`RemainingTableCount`
|`int`
|The number of tables that the snapshot has yet to copy.

|`SnapshotRunning`
|`boolean`
|Whether the snapshot was started.

|`SnapshotAborted`
|`boolean`
|Whether the snapshot was aborted.

|`SnapshotCompleted`
|`boolean`
|Whether the snapshot completed.

|`SnapshotDurationInSeconds`
|`long`
|The total number of seconds that the snapshot has taken so far, even if not complete.

|`RowsScanned`
|`Map<String, Long>`
|Map containing the number of rows scanned for each table in the snapshot. Tables are incrementally added to the Map during processing. Updates every 10,000 rows scanned and upon completing a table.

|=======================

[[monitoring-streaming]]
[[streaming-metrics]]
==== Streaming Metrics

> ===== *MBean: debezium.sql_server:type=connector-metrics,context=streaming,server=_<database.server.name>_*

Reviewer: sql_server -> db2

Author: Done

> [cols="30%a,10%a,60%a",width=100,options="header,footer",role="table table-bordered table-striped"]

Reviewer: @Naros, does this need updating as per your recent change? This may have overlapped with this PR which I only got to review now.

|`MonitoredTables`
|`string[]`
|The list of tables that are monitored by the connector.

|`QueueTotalCapcity`
|`int`
|The length of the queue used to pass events between the streamer and the main Kafka Connect loop.

|`QueueRemainingCapcity`
|`int`
|The free capacity of the queue used to pass events between the streamer and the main Kafka Connect loop.

|`Connected`
|`boolean`
|Flag that denotes whether the connector is currently connected to the database server.

|`MilliSecondsBehindSource`
|`long`
|The number of milliseconds between the last change event's timestamp and the connector processing it. The values will incorporate any differences between the clocks on the machines where the database server and the Debezium connector are running.

|`NumberOfCommittedTransactions`
|`long`
|The number of processed transactions that were committed.

|`SourceEventPosition`
|`map<string, string>`
|The coordinates of the last received event.

|`LastTransactionId`
|`string`
|Transaction identifier of the last processed transaction.

|=======================

[[monitoring-schema-history]]
[[schema-history-metrics]]
==== Schema History Metrics

> ===== *MBean: debezium.sql_server:type=connector-metrics,context=schema-history,server=_<database.server.name>_*

Reviewer: sql_server -> db2

Author: Done

|`ChangesRecovered`
|`long`
|The number of changes that were read during recovery phase.

|`ChangesApplied`
|`long`
|The total number of schema changes applie during recovery and runtime.

|`MilliSecondsSinceLastRecoveredChange`
|`long`
|The number of milliseconds that elapsed since the last change was recovered from the history store.

|`MilliSecondsSinceLastAppliedChange`
|`long`
|The number of milliseconds that elapsed since the last change was applied.

|`LastRecoveredChange`
|`string`
|The string representation of the last change recovered from the history store.

|`LastAppliedChange`
|`string`
|The string representation of the last applied change.

|=======================

[[connector-properties]]
=== Connector properties

The following configuration properties are _required_ unless a default value is available.

[cols="35%a,10%a,55%a",options="header,footer",role="table table-bordered table-striped code-wordbreak-col"]
|=======================
|Property
|Default
|Description

|`name`
|
|Unique name for the connector. Attempting to register again with the same name will fail. (This property is required by all Kafka Connect connectors.)

|`connector.class`
|

> |The name of the Java class for the connector. Always use a value of `io.debezium.connector.sqlserver.SqlServerConnector` for the Db2 connector.

Reviewer: ```suggestion
|The name of the Java class for the connector. Always use a value of `io.debezium.connector.db2.Db2Connector` for the Db2 connector.
```

Author: Done

|An integer value that specifies the maximum amount of time (in milliseconds) to wait to obtain table locks when performing a snapshot. If table locks cannot be acquired in this time interval, the snapshot will fail (also see link:#snapshots[snapshots]). +
When set to `0` the connector will fail immediately when it cannot obtain the lock. Value `-1` indicates infinite waiting.

|`snapshot.select.statement.overrides`
|
|Controls which rows from tables will be included in snapshot. +
This property contains a comma-separated list of fully-qualified tables _(SCHEMA_NAME.TABLE_NAME)_. Select statements for the individual tables are specified in further configuration properties, one for each table, identified by the id `snapshot.select.statement.overrides.[SCHEMA_NAME].[TABLE_NAME]`. The value of those properties is the SELECT statement to use when retrieving data from the specific table during snapshotting. _A possible use case for large append-only tables is setting a specific point where to start (resume) snapshotting, in case a previous snapshotting was interrupted._ +
*Note*: This setting has impact on snapshots only. Events captured during log reading are not affected by it.

> |`source.struct.version`

Reviewer: That option isn't existing for that connector.

Author: Is this connector specific or handled by the core (we thought the latter) ?

Author: Done

|v2
|Schema version for the `source` block in CDC events; Debezium 0.10 introduced a few breaking +
changes to the structure of the `source` block in order to unify the exposed structure across
all the connectors. +
By setting this option to `v1` the structure used in earlier versions can be produced.
Note that this setting is not recommended and is planned for removal in a future Debezium version.

|`sanitize.field.names`
|`true` when connector configuration explicitly specifies the `key.converter` or `value.converter` parameters to use Avro, otherwise defaults to `false`.
|Whether field names will be sanitized to adhere to Avro naming requirements.
See xref:configuration/avro.adoc#names[Avro naming] for more details.

> |`database.server.timezone`

Reviewer: Does this option actually exist for this connector?

Author: Done
